### PR TITLE
save learned mag bias per sensor (Multi-EKF support)

### PR DESCRIPTION
**Pull request by dagar**
- move the "mag calibration" code from ekf2 to sensors/vehicle_magnetometer
 - handle saving the mag bias per sensor (across all estimator instances using that mag)
 - currently this is saving back to the actual mag calibration (CAL_MAGn_OFF{X,Y,Z}), but could be separate parameters
    - this could still work in conjugation with an underlying factory calibration or we could separately maintain the bias
 - ekf2 reset mag mag bias on any magnetometer or calibration change (new ecl/EKF resetMagBias() should be reviewed). https://github.com/PX4/PX4-ECL/blob/a8e0e82858318042500f96f5ee882c638f201426/EKF/ekf_helper.cpp#L878-L892

### CODE REVIEW

**Comment by dagar** [comment]
There are at least 2 questions that need to be settled here.
1. Where do we save?
   a) update existing CAL_MAGn_OFF{X, Y, Z}
       - pro: doesn't require additional parameters or special logic to clear the EKF2 mag bias after a new calibration 
       - con: overwrites the existing calibration that we may wish to preserve (for comparison over time?)
   b) separately store bias per sensor ID (eg EKF2_MAGBIAS_{X,Y,Z} -> SENS_MAGn_BIAS_{X,Y,Z})
2. How best to handle multiple estimators using the same magnetometer?
  - the current change simply averages each new valid mag bias update with the existing (across all available instances)

File: src/lib/sensor_calibration/Magnetometer.hpp [hunk 1]
```diff
@@ -85,6 +85,12 @@ class Magnetometer
 		return _rotation * (_scale * ((data + _power * _power_compensation) - _offset));
 	}
 
+	// Compute sensor offset from bias (board frame)
+	matrix::Vector3f BiasCorrectedSensorOffset(const matrix::Vector3f &bias) const
+	{
+		return _scale.I() * _rotation.I() * bias + _offset;
```
Comment (priseborough) [inline src/lib/sensor_calibration/Magnetometer.hpp:91]: Is this equation correct? I think it is, but can't find documentation for which frame of reference the class variables _scale, _rotation and _offset are defined in and the order that rotations and offsets should be applied. This function implies that bias and _offset are in board frame, _rotation rotates from sensor frame to board frame and _scale removes scale and soft iron errors in sensor frame after removal of offsets.

Comment (dagar) [reply src/lib/sensor_calibration/Magnetometer.hpp:91]: This `Magnetometer` class (src/lib/sensor_calibration/Magnetometer.hpp) is where I've collected all the corrections applied to raw sensor data that's been ripped out of the drivers and sensors module. All sensors now publish raw data (sensor_accel/gyro/mag) and everything is applied downstream. The only exception is for internal sensors that are started by the board with a particular rotation. 

Ultimately this is what's applied to each raw sample. https://github.com/PX4/PX4-Autopilot/blob/64240010bb67b75f880a4972189dadd988055b17/src/lib/sensor_calibration/Magnetometer.hpp#L83-L86

So yes, we should really do a pass to thoroughly document everything.

File: src/modules/ekf2/EKF2.cpp [hunk 2]
```diff
@@ -1376,62 +1340,9 @@ void EKF2::UpdateRangeSample(ekf2_timestamps_s &ekf2_timestamps)
...
-		if (_last_magcal_us == 0) {
-			_last_magcal_us = timestamp;
-
-		} else {
-			_total_cal_time_us += timestamp - _last_magcal_us;
-			_last_magcal_us = timestamp;
-		}
-
-		// Start checking mag bias estimates when we have accumulated sufficient calibration time
-		if (_total_cal_time_us > 30_s) {
```
Comment (priseborough) [inline src/modules/ekf2/EKF2.cpp:1394]: We have effectively removed this 30 second requirement. This is probably OK given the checks on variances, but will need to be checked with a series of short up and down flights including some yaw rotations.

File: src/modules/sensors/vehicle_magnetometer/VehicleMagnetometer.cpp [hunk 3]
```diff
@@ -141,18 +144,140 @@ void VehicleMagnetometer::ParametersUpdate(bool force)
...
+				if (_mag_cal_valid[i].get_state()) {
+					// find corresponding mag calibration
+					for (int mag_index = 0; mag_index < MAX_SENSOR_COUNT; mag_index++) {
+						if (_calibration[mag_index].device_id() == estimator_sensor_bias.mag_device_id) {
+
+							const Vector3f mag_offset{_calibration[mag_index].BiasCorrectedSensorOffset(bias)};
+
+							// store best state to save back to parameters later, averaged across valid estimates
+							if (_mag_cal_available[mag_index]) {
+								_mag_cal_offset[mag_index] = (_mag_cal_offset[mag_index] + mag_offset) * 0.5f;
```
Comment (priseborough) [inline src/modules/sensors/vehicle_magnetometer/VehicleMagnetometer.cpp:182]: I can't find where this average of the two last valid estimates comes from. It is probably unnecessary given the way that the weighting is applied on line 221

Comment (priseborough) [reply src/modules/sensors/vehicle_magnetometer/VehicleMagnetometer.cpp:182]: Is this being used to accomodate updates from more than one EKF using the same magnetometer?

File: src/modules/ekf2/EKF2.cpp (hunk 2 above)
Comment (dagar) [reply src/modules/ekf2/EKF2.cpp:1485]: It's now handled by a hysteresis. https://github.com/PX4/PX4-Autopilot/pull/16417/files#diff-82b382c1f89f118ca86cd0ad5765be02769ce61703f41079537eee1635aa34c5R111

**Comment by dagar** [comment]
TODO:
 - store all valid mag biases from flight (across all available estimators)
 - use all to incrementally update the mag cal offset, each using the variance weight from flight

File: src/modules/ekf2/EKF2.cpp (hunk 2 above)
Comment (dagar) [reply src/modules/ekf2/EKF2.cpp:1485]: Nevermind, I kept it in ekf2 to properly handle 30 seconds of accumulated 3D mag fusion time.

**Comment by dagar** [comment]
This is working and ready for further review.

Tested in SITL with multi-EKF enabled (3 IMUs, 2 mags). I manually added a bias of 0.1 Gauss to the X axis of the first mag, and 0.2 Gauss to the Y axis of the 2nd.

Flights
1.  https://logs.px4.io/plot_app?log=272cf201-1205-4196-b64b-3053515a2cd1
2. https://logs.px4.io/plot_app?log=42255994-d39a-401e-86d1-714602d6e9b0
3. https://logs.px4.io/plot_app?log=ea42d3eb-9478-474e-95e7-65b401d88870
4. https://logs.px4.io/plot_app?log=2b1e5a5f-f494-42dc-a15f-63cc5e4c315e


Here's the relevant mag calibration by the end of the 4th flight.

``` Console
x + CAL_MAG0_ID [67,114] : 197388
x + CAL_MAG0_XOFF [72,119] : 0.0963 (should be 0.1)
x + CAL_MAG0_YOFF [76,123] : 0.0003
x + CAL_MAG0_ZOFF [80,127] : -0.0027

x + CAL_MAG1_ID [82,129] : 197644
x + CAL_MAG1_XOFF [87,134] : 0.0001
x + CAL_MAG1_YOFF [91,138] : 0.1931 (should be 0.2)
x + CAL_MAG1_ZOFF [95,142] : -0.0002
```


For example here's the output from log 4 on landing where you can see the learned offsets being updated recursively.

``` Console
INFO  [commander] Disarmed by landing
INFO  [vehicle_magnetometer] Mag 0 (197388) est. offset 0 committed: [ 0.083  0.000 -0.004] -> [ 0.090  0.000 -0.003] (full [ 0.100  0.000 -0.002])
INFO  [vehicle_magnetometer] Mag 0 (197388) est. offset 1 committed: [ 0.090  0.000 -0.003] -> [ 0.094  0.000 -0.003] (full [ 0.100  0.000 -0.002])
INFO  [vehicle_magnetometer] Mag 0 (197388) est. offset 2 committed: [ 0.094  0.000 -0.003] -> [ 0.096  0.000 -0.003] (full [ 0.100  0.000 -0.002])
INFO  [vehicle_magnetometer] Mag 1 (197644) est. offset 3 committed: [-0.000  0.168 -0.000] -> [-0.000  0.181 -0.000] (full [ 0.000  0.200 -0.000])
INFO  [vehicle_magnetometer] Mag 1 (197644) est. offset 4 committed: [-0.000  0.181 -0.000] -> [ 0.000  0.189 -0.000] (full [ 0.000  0.200 -0.000])
INFO  [vehicle_magnetometer] Mag 1 (197644) est. offset 5 committed: [ 0.000  0.189 -0.000] -> [ 0.000  0.193 -0.000] (full [ 0.000  0.200 -0.000])
```

**Comment by priseborough** [comment]
I've created a patch which implements a Kalman filter method to incorporate the multiple offset estimates. This is effectively the same as what we had before but with the important distinction that the state variance is updated after each fusion cycle. See https://github.com/priseborough/Firmware/commit/bd68772dfe6bd4ff0a2cfa7fb87e15bb9dcb0ce6

**Comment by dagar** [comment]
Thanks Paul, branch updated with your change. I'll test again.

**Comment by priseborough** [comment]
I've flight tested the patch and confirmed via flight test that the bias is corrected in the right direction.

**Comment by priseborough** [comment]
I'm testing on a F7 board with dual GPS/mag units and EKF2_MULTI_IMU and EKF2_MULTI_MAG set to 2 and SENS_IMU_MODE and SENS_MAG_MODE set to 0 and am seeing a more instance changes than expected:

Screen Shot 2021-01-07 at 12 44 23 pm

This was a test starting with a deliberately added 100mGauss change to the MAG0 Y offset, so will do a sensor cal and check again.

**Comment by priseborough** [comment]
Checking again with a clean build and all submodules updated following a fresh magnetometer calibration and then with a prior 100mGauss offset to the Y axis biases gives:

Following recalibration fo the magnetometers:

Screen Shot 2021-01-07 at 1 21 09 pm

With change of 100mGauss to the Y axis offsets:

https://github.com/priseborough/Firmware/commit/bd68772dfe6bd4ff0a2cfa7fb87e15bb9dcb0ce6

The rapid switching is still there.

**Comment by priseborough** [comment]
Screen Shot 2021-01-07 at 1 47 00 pm

The lane switching is being caused by transient bad conditioned X-axis mag fusion that occur during high angular rates. I will grab a replayable log with one EKf enabled and debug using replay.

**Comment by dagar** [comment]
The new `estimator_status_flags` message should work a bit better for catching these brief faults because it publishes only on change (or at 1 Hz) and is fully logged.

**Comment by priseborough** [comment]
I found the cause of the mag fusion x-axis fault flag. It happens when the _accel_bias_inhibit is true and the delta velocity bias state variances and covariances are being set to 0 in the covariance prediction. This results in the checkAndFixCovarianceUpdate function returning false because P(i,i) is 0 and KHP(i,i) is a very small positive number (eg 1E-16). I will fix this in a separate PR for ecl.

The mag offset parameter updating changes introduced by this PR are working as expected.

**Review by priseborough — APPROVED** [review]

**Pull request merged**